### ISSUE TITLE
Check current API surface

### ISSUE THREAD

**Issue by orottier** [orottier/web-audio-api-rs#76]
API surface: `cargo doc --lib --open`

Check for

- [x] consistency
- [x] internals accidentally marked `pub`
- [x] commit to the current `mod` layout?
- [x] commit to the location of the `Error`s and constants?
- [x] happy with the `BaseAudioContext` and `AsBaseAudioContext`?
- [x] `AudioContext::new(AudioContextOptions)` (not wrapped in an Option)
- [x] remove `Scheduler` and `Controller` from public interface, change traits (AudioScheduledSourceNode etc)
- [x] Check all node constructor options
- [x] Check all node methods
- [x] remove public spatial mod
- [x] mark functions that should return a Promise but currently don't `_sync`
- [x] rename `AudioContextOptions.channels -> number_of_channels`
- [x] create_delay and create_channel_(splitter/merger) take optional argument, but we make them required. Same as create_periodic_wave but here we take in an opt struct. 
- [x] Can we drop  AudioNode's  `channel_config_raw` and `channel_config_cloned`?
- [x] Consistency in AudioNode optional arg: `disconnect() , disconnect_from(node)`
- [x] Remove AudioNodeId from public interface
- [x] stream node constructors only on online AudioContext

**Comment by b-ma** [comment]
> commit to the current mod layout?

- My very personal taste from a user point of view would be to have everything flat. As the public API is not that huge and there is possibility of naming collision (in JS everything is basically exposed to `window`), the namespacing feels more like a burden people need to learn than something that helps. 
- However, I also think that it would be nice to have some `extra` (or whatever) namespace that would contain all structs that are not _per se_ part of the API, but are really nice to have at hand (e.g. `Microphone`, `MediaElement`, etc.) and without which some nodes of the API do not make much sense.

> happy with the BaseAudioContext and AsBaseAudioContext?

Not really :), I think the `AsBaseAudioContext` is really weird from a user point of view and that it would be far more simple to understand renamed to `BaseAudioContext`, e.g.: I can understand that I have to `import::context::{AudioContext, BaseAudioContext}`, but I really don't get why i should `import::context::{AudioContext, AsBaseAudioContext}` which is not even described in the spec. 

From an implementation perspective `AsBaseAudioContext` really makes sense, but from a user perspective it really does not.

**Comment by b-ma** [comment]
Also, just to let you know (maybe that may shade some other light on these questions) that I restarted to work on the Nodejs bindings for the lib: https://github.com/b-ma/node-web-audio-api-rs 

My idea at some point (as most of the code is basically `unwrap` / `cast` / `wrap`) would be to generate most of the code from the spec IDL (https://webaudio.github.io/web-audio-api/#idl-index), so I guess this could help to spot inconsistencies in the API. Maybe also, in a more mid/long term perspective, this could allow to test the lib against the existing Web Audio test suite https://github.com/web-platform-tests/wpt/tree/master/webaudio without rewriting everything in pure rust.

_Aside_: it would be really helpful to have your insights on the rust code here (especially with the `AudioParam` bindings as I spent literally hours fighting with the borrow checker here... and I'm really not sure my final solution is very good). 
Also note that I managed to port the `many_oscillators_with_env` example and that it works pretty well: CPU looks stable, and memory too (it grows up to 50Mo which seems a lot, but I guess (TBC) this rather comes from how Node works internally than from the bindings, as once ~50Mo is reached it seems to stay stable there), so that's very nice!

**Comment by orottier** [comment]
Ah it's really cool that you're working on the JS bindings! Testing against the official test suite was something I had in mind for a long time but I was dreading the amount of work needed to provide the bindings :)
I will have a look at your repo, sure. If you have any specific questions we can address them there.

> commit to the current mod layout?
I think we should strike a balance between the rust and JS preferences. To be completely flat is a bit unheard of in (larger) rust projects. Surely we can get rid of `buffer`, `control` and `spatial`.
I would like to keep `render` as a mod that you only need to have a look at if you are implementing AudioNode/Render yourself. We can then move AudioParamOptions there, and move the rest of param.rs to the global namespace. Let's see where that takes us and iterate from there onwards

> From an implementation perspective AsBaseAudioContext really makes sense, but from a user perspective it really does not.
Yeah, agree... Let's get rid of the trait and just copy/paste the functions in the AudioContext and OfflineAudioContext

**Comment by b-ma** [comment]
> happy with the BaseAudioContext and AsBaseAudioContext?

In the same vein, I think we could question the `ScheduledAudioSourceNode` trait, because as it's already not implemented in the `AudioBufferSourceNode` (because of the `start_at_with_offset`, etc.), we fall in a strange situation where we have to import the trait when using an `OscillatorNode` or a `ConstantSourceNode` but not a `ScheduledAudioSourceNode`. It seems to me that it makes sens from an OOP perspective but just make things more complicated here.

**Comment by b-ma** [comment]
Another thing that feels strange to me is the `disconnect` / `disconnect_all` (just discovered in your new example), that's really a point of view but, intuitively I would rather wait for the inverse:
- `disconnect()` -> disconnect from all the node
- `disconnect_from(node)` -> disconnect from a given node

I also have the impression this would be more consistent which how optional params are handled in the rest of the API (e.g. `start()` / `start_at(time)`)

**Comment by tasogare3710** [comment]
Hi,

@orottier
> Yeah, agree... Let's get rid of the trait and just copy/paste the functions in the AudioContext and OfflineAudioContext

Why don't you use delegation(not GoF)?


- https://lib.rs/search?q=delegate
- https://internals.rust-lang.org/search?q=Delegation

**Comment by orottier** [comment]
Thanks for the suggestion @tasogare3710 , I did not know about that whole category of crates!

I have two questions though,
- will it also delegate the doc strings?
- if we need to delegate every method, how can we enforce we keep them all in sync? (e.g. if we add create_convolver to the base context, we must not forget to add the new method to both AudioContext and OfflineAudiocontext)

A macro would not have these limitations.

In any case, feel free to whip up a PR using delegation!

**Comment by tasogare3710** [comment]
Sorry delay reply, @orottier.

In this crate, we need to define a delegate method and implement a trait with a method of the same name, so we need to write a macro myself.

> will it also delegate the doc strings?

rustdoc does not have a way to control document inheritance like javadoc's inheritdoc tag, use a macro.

> if we need to delegate every method, how can we enforce we keep them all in sync? (e.g. if we add create_convolver to the base context, we must not forget to add the new method to both AudioContext and OfflineAudiocontext)

Since the methods to be delegated are known in advance, they can be automated using the proc macro ...I think it is.
Basically, methods that are called indirectly are target of a delegatation.

**Comment by tasogare3710** [comment]
I'm working on it and would like some advice.

https://github.com/tasogare3710/web-audio-api-rs/tree/remove_BaseAudioContext_and_hide_AudioContext


 - Do I need to private the BaseAudioContext?
 - Is there anything else I should do?

**Comment by orottier** [comment]
Let's move that discussion to #99

**Comment by b-ma** [comment]
Are we ok with the `AudioContextOptions` ?

```rs
pub struct AudioContextOptions {
    /// Identify the type of playback, which affects
    /// tradeoffs between audio output latency and power consumption
    pub latency_hint: Option<AudioContextLatencyCategory>,
    /// Sample rate of the audio Context and audio output hardware
    pub sample_rate: Option<u32>,
    /// Number of output channels of destination node and audio output hardware
    pub channels: Option<u16>,
}
```

**Comment by b-ma** [comment]
Yup I'm finally quite agreeing that having a `Default::default()` is better `Option` here. However I'm not sure we follow the same rules as for the nodes (see #103), the IDL being:

```
dictionary AudioContextOptions {
  (AudioContextLatencyCategory or double) latencyHint = "interactive";
  float sampleRate;
};
```

- I don't understand why `latency_hint: Option<AudioContextLatencyCategory>` is an option here?
- `sampleRate` to `None` by default meaning default system sample rate seems very logical
- do we have to keep `channels` here (should live in `AudioListener` I guess) because of the way the `cpal` stream is created?

_Aside_: I just discovered my proposal there:
> disconnect() -> disconnect from all nodes
> disconnect_from(node) -> disconnect from a given node 

Do you agree or do you prefer the way it is now?

**Comment by orottier** [comment]
I am working on `AudioContext::new(AudioContextOptions)` now

**Comment by orottier** [comment]
I think #134 is the last PR I wanted to submit for this ticket.
I'm quite happy with the mod layout (using `media` instead of `extra`) and I think the other issue around required arguments is not too important.
Agree @b-ma ?

**Comment by b-ma** [comment]
Hey, 

Yup, looks pretty good to me indeed,

- for `extra` vs `media`, I personally prefer `extra` but that's not a strong opinion neither, much more a matter of documentation than anything else I think
- for the optional args, I agree with you it doesn't seem very interesting to wrap things into `Option`s, everything being required seems consistent to me (actually I wonder if we should make it a rule and ignore `optional` params everywhere as it doesn't make sens in Rust as it does in JS, maybe we are already mostly there...)

At rapid sight, my only concern about the context's factory methods is for the `create_periodic_wave`: now we take a `PeriodicWaveOptions` which seems very logical to me, but the spec is:

```
  PeriodicWave createPeriodicWave (sequence<float> real,
                                   sequence<float> imag,
                                   optional PeriodicWaveConstraints constraints = {});
```

The spec is much less clean and looks like an historical bloat in my opinion, so I don't really know what we should do (f we stick to the current API, we should at least add a comment so we don't re-ask this question again and again)

Also, I finally wonder if we should put the `stream` related factory method only in the online context as per spec (cf. https://webaudio.github.io/web-audio-api/#AudioContext):
- maybe implementing such thing `offline` will only add huge constraints and create problems we don't really want to deal with
- we could add them `pub(crate)` to the offline context so we can use them in tests (and possibly expose them later when we are more confident about the different paradigms and if everything works as expected)

**Comment by b-ma** [comment]
> Regarding the optional args, we cannot drop all of them, only when they have a Default implementation.
> For example AudioBufferSourceOptions and WaveShaperOptions have an optional field that we cannot ignore. For the rest I think we are all set!

Yes you right, didn't think about these ones. I guess things start to be very consistent now :) that's nice !

**Comment by b-ma** [comment]
really really nice :)

**Comment by b-ma** [comment]
Just to note, I have seen some weirdness with the Panner node when I tried to generate the node bindings. 
Didn't really dig into the problem, so that might be on the bindings side I don't really know, but I will try to have a closer look on that soon
